src/midi/midi_fluidsynth.cpp

```diff
@@ -426,14 +426,19 @@ uint16_t MidiHandlerFluidsynth::GetRemainingFrames()
 // released in the ring allowing FluidSynth to renderer the next buffer.
 void MidiHandlerFluidsynth::Render()
 {
-	std::vector<float> render_buffer(FRAMES_PER_BUFFER * 2); // L & R channels
+	constexpr auto SAMPLES_PER_BUFFER = FRAMES_PER_BUFFER * 2; // L & R
+	std::vector<float> render_buffer(SAMPLES_PER_BUFFER);
+	std::vector<int16_t> playable_buffer(SAMPLES_PER_BUFFER);
+
 	while (keep_rendering) {
 		fluid_synth_write_float(synth.get(), FRAMES_PER_BUFFER,
 		                        render_buffer.data(), 0, 2, render_buffer.data(), 1, 2);
-		auto out = soft_limiter.Process(render_buffer, FRAMES_PER_BUFFER);
-		ring.wait_enqueue(out);
+		soft_limiter.Process(render_buffer, FRAMES_PER_BUFFER,
+		                     playable_buffer);
+		ring.wait_enqueue(playable_buffer);
```

Comment:
I might've been wrong in suggesting to avoid creating temporary objects here…

Which `wait_enqueue` is being called in here?

`void wait_enqueue(T const& item)` (in which case, according to rwqueue documentation, the vector is copied, so we get limited benefit from reusing the object. <- I hope this is the case. (It's not necessarily bad that it's copied - copying larger vector like this probably happens via fast memcpy copying whole memory blocks at once).

or `void wait_enqueue(T&& item)` (in which case our locally-created `playable_buffer` is moved out, so starting with 2nd iteration we're writing over random memory while filling it in inside `Process`)?

*Side note: I am starting to think we don't need atomic rwqueue at all - we need much simpler structure implementing double or tripple-buffering. Perhaps it already works like that, just with unnecessary steps along the way.*

Reply:
> I might've been wrong in suggesting to avoid creating temporary objects here…

The PR is a three-step transition; each one is compilable. In this first commit, we're just adjusting soft limiter (in the commit comment). 

> Which wait_enqueue is being called in here?

Correct, this is still the copy call being used in the first commit; I just changing one thing at a time, and keeping the code buildable and testable at each step.

Reply:
> Side note: I am starting to think we don't need atomic rwqueue at all - we need much simpler structure implementing double or tripple-buffering. Perhaps it already works like that, just with unnecessary steps along the way.

The rw-queue's ability to let us move objects in and out of a simple queue in two places of code (that are actually being run in separate threads) belies the work it's doing under the hood to give us such an easier to understand workflow up in our code.

  It's easy to forget it's even multithreaded (I'm not saying you are), and the code makes it easy to envision using different container or queue in it's place, which is testament to how easy it's made the code to read. 

By the 3rd commit, you will see the batch of 8 allocated buffers being rotated in and out of two queues, all using moves.  It's a sweet as it gets :-) this would normally be significantly harder without the rw-queue. 

But as for experimenting with other structures or buffering techques - please go for it! By the 3rd commitw, it's quite readable exactly how the buffers are rotating so should be easy to try new ideas :-) 👍